enhancements/installer/pre-flight-validations.md

```diff
@@ -0,0 +1,194 @@
+succeed. This goal is relatively easy to implement for public cloud platforms
+where we can make assumptions about services being available, or performance
+meeting requirements, however this is not the case with private clouds where
+each cloud is unique.
+
+It is currently not possible to tell with confidence if an installation will be
+successful or not for the following platforms:
+- BareMetal
+- OpenStack
+- oVirt
+- vSphere
+
+For this purpose, we propose to implement a framework allowing the installer to
+run pre-flight validations in order to certify that all pre-requisites are met
+to successfully install OpenShift in the selected environment.
+
+## Motivation
+
+This section is for explicitly listing the motivation, goals and non-goals of
+this proposal. Describe why the change is important and the benefits to users.
+
+### Goals
+
+By having an automated way to identify potential issues early, before the
+deployment even started, we want to reduce wasted time and resources, customer
+escalations, and improve the perception of OpenShift deployments on-premise.
+
+The installer already performs some validations:
+- checks all required fields are set and that the data is in the right format
+- some basic validation, such as networks do not overlap
+
+However, this doesn't check that the environment is suitable to install OpenShift:
+- pull secret is valid to fetch the container images
+- the tenant has adequate quota and the flavors' specifications are within the
+  recommended ranges.
+- for user-provided networks, check the subnets have a DHCP server and valid
+  DNS to reach the cloud's enpoints
+- necessary cloud services are available
+- storage performance
+
+### Non-Goals
+
+Implementing every useful validation we can think of is out-of-scope:
+- The potential number of validations is huge
+- We expect more validations to be added over time as new issues are
+  discovered.
+
+## Proposal
+
+This is where we get down to the nitty gritty of what the proposal actually is.
+
+### User Stories
+
+#### On-premise deployments
+
+As an administrator of an OpenShift cluster, I would like to verify that my
+OpenStack cloud meets all the performance, service and networking requirements
+necessary for a successful deployment.
+
+#### CI debugging
+
+As an OpenShift developer, I would like to know if the CI flakes might be
+caused by transient environmental issues.
+
+### Implementation Details/Notes/Constraints [optional]
+
+This should allow for checks common to all platforms and per-platform checks.
+
+We do not envision the need for the users to write additional validations. As
+a consequence, the validations do not need to be loaded on startup and will be
+compiled into the `openshift-install` binary. This may be revisited later.
+
+#### Pre-provision a node
+
+Node with the master flavor on the user provisioned network:
+- pull container images
+- validate networking and cloud connectivity
+- run fio
```

Comment:
What's fio?

Reply:
I guess https://github.com/axboe/fio ?

Reply:
Yes that's the one. It's a benchmarking tool for storage, and is the recommended tool to check that storage performance meets the requirements for etcd:
https://github.com/etcd-io/etcd/blob/master/Documentation/faq.md#system-requirements